### PR TITLE
Return the user-desired HTTP method, if given.

### PR DESCRIPTION
**Pull request by elpete**
Return the user-desired HTTP method, if given.  Otherwise, return the CGI.REQUEST_METHOD, as usual.

See https://laravel.com/docs/5.2/routing#form-method-spoofing

This will introduce no breaking issues while making it possible for normal forms to follow RESTful routing guidelines.

| HTTP Verb      | Path            | Handler@Action | Used for                                     |
| -------------- | --------------- | -------------- | -------------------------------------------- |
| GET            | `/posts`          | posts@index    | Displays a list of all posts                 |
| GET            | `/posts/new`      | posts@new      | Returns an HTML form for creating a new post |
| POST           | `/posts`          | posts@create   | Creates a new post                           |
| GET            | `/posts/:id`      | posts@show     | Displays a specific post                     |
| GET            | `/posts/:id/edit` | posts@edit     | Returns an HTML form for editing a post      |
| PATCH/PUT      | `/posts/:id`      | posts@update   | Updates a specific post                      |
| DELETE         | `/posts/:id`      | posts@destroy  | Deletes a specific post                      |

```cfml
<form method="POST" action="#event.buildLink( 'posts.#post.getId()#' )#">
    <input type="hidden" name="_method" value="DELETE" />
   <button type="submit">Delete</button>
</form>

<!--- Instead of --->

<form method="POST" action="#event.buildLink( 'posts.#post.getId()#.delete' )#">
   <button type="submit">Delete</button>
</form>
```

The HTMLHelper has been updated as well:

```cfml
#html.startForm( action = "posts.#prc.post.getId()#", method="DELETE" )#
    #html.submitButton( name = "Delete", class = "btn btn-danger" )#
#html.endForm()#

<!--- will output --->

<form action="http://localhost/posts/1" method="POST">
    <input type="hidden" name="_method" value="DELETE" />
    <input type="submit" name="Delete" value="Submit" class="btn btn-danger" id="Delete"/>
</form>
```

### CODE REVIEW

**Comment by lmajano** [comment]
This looks great. Would you mind adding the docs for this in the ColdBox docs ?

**Comment by lmajano** [comment]
Also a ticket please

**Comment by elpete** [comment]
[COLDBOX-525](https://ortussolutions.atlassian.net/browse/COLDBOX-525)

File: system/core/dynamic/HTMLHelper.cfc [hunk 1]
```diff
@@ -573,8 +573,17 @@ Description :
...
-			var formBuffer	= createObject( "java", "java.lang.StringBuilder" ).init( "<form" );
-			var event 		= controller.getRequestService().getContext();
+			var formBuffer	  = createObject( "java", "java.lang.StringBuilder" ).init( "<form" );
+			var event         = controller.getRequestService().getContext();
+			var desiredMethod = '';
+
+			// Browsers can't support all the HTTP verbs, so if we passed in something
+			// besides GET or POST, we'll default to POST and save off
+			// the desired method to spoof later.
+			if (arguments.method != "GET" && arguments.method != "POST") {
```
Comment (lmajano) [inline system/core/dynamic/HTMLHelper.cfc]: Hi Pete, sorry to be picky, but we have started to be stricter with our syntax guidelines.  Especially on spacing.  Can you convert this to our white space rules:

```
if( arguments.method != "GET" && arguments.method != "POST" ){
```

File: system/core/dynamic/HTMLHelper.cfc [hunk 2]
```diff
@@ -604,6 +613,12 @@ Description :
 			// create tag
 			flattenAttributes( arguments, "noBaseURL,ssl,multipart", formBuffer )
 				.append( ">" );
+				
+			// If we wanted to use PUT, PATCH, or DELETE, spoof the HTTP method
+			// by including a hidden field in the form that ColdBox will look for.
+			if (len( desiredMethod )) {
```
Comment (lmajano) [inline system/core/dynamic/HTMLHelper.cfc]: Spacing again
`if( len( desiredMethod ) ){`

File: system/core/dynamic/HTMLHelper.cfc [hunk 3]
```diff
@@ -604,6 +613,12 @@ Description :
...
+				formBuffer.append( "<input type=""hidden"" name=""_method"" value=""#desiredMethod#"" />");
```
Comment (lmajano) [inline system/core/dynamic/HTMLHelper.cfc]: `formBuffer.append( "<input type=""hidden"" name=""_method"" value=""#desiredMethod#"" />" );`

**Comment by elpete** [comment]
Looks like that failure was due to pulling down the key for CommandBox.  I'd re-run it if I could.

**Comment by elpete** [comment]
Documented at https://github.com/ColdBox/coldbox-docs/pull/21

**Pull request merged**